source/io/module_javaScriptInvoke.js

```diff
@@ -358,30 +372,52 @@
                 newErrorCode = ERRORS.kNIUnableToFindFunctionForJavaScriptInvoke.CODE;
                 newErrorSource = ERRORS.kNIUnableToFindFunctionForJavaScriptInvoke.MESSAGE + '\'' + functionName + '\'.';
                 Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                Module.eggShell.setOccurrenceAsync(occurrencePointer);
                 return;
             }
 
-            var context = undefined;
+            var returnTypeName = getParameterTypeString(returnPointer, 0);
+            var returnValuePointer = undefined;
+            if (returnTypeName !== 'StaticTypeAndData') {
+                // User doesn't want return value. We're passing '*' for the return in VIA code, we get StaticTypeAndData
+                returnValuePointer = JavaScriptInvoke_GetParameterPointer(returnPointer, 0);
+            }
+            if (returnTypeName === 'Array') {
+                returnTypeName += getArrayElementTypeString(returnValuePointer);
+            }
+
+            var asyncFlag = false;
+            var generateContext = function () {
+                var completionCallbackRetrieved = false;
+                var context = {};
+                context.getCompletionCallback = function () {
+                    if (completionCallbackRetrieved === true) {
+                        throw new Error('The completion callback was retrieved more than once.');
+                    }
+                    asyncFlag = true;
+                    completionCallbackRetrieved = true;
+                    return generateCompletionCallback(occurrencePointer, functionName, returnTypeName, returnValuePointer, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                };
+                return context;
+            };
+
             var returnValue = undefined;
+            var context = generateContext();
             try {
                 returnValue = functionToCall.apply(context, parameters);
             } catch (ex) {
                 newErrorStatus = true;
                 newErrorCode = ERRORS.kNIUnableToInvokeAJavaScriptFunction.CODE;
                 newErrorSource = Module.coreHelpers.formatMessageWithException(ERRORS.kNIUnableToInvokeAJavaScriptFunction.MESSAGE + '\'' + functionName + '\'.', ex);
                 Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                Module.eggShell.setOccurrenceAsync(occurrencePointer);
```

Comment:
Hmm so suppose hypothetically during functionToCall.apply(context, parameters) that the user has a function like:
```javascript
function () {
   var cb = this.generateCompletionCallback();
   document.getElementById('mybutton').addEventListener('click', cb.bind(undefined, 'hello world'));
   throw new Error('woops');
}
```

So here we capture the Exception and finish the javaScriptInvoke call and wire out error terminals. But the user can click a button on the page, which will trigger the completion callback, and start writing things into Vireo memory and corrupt it.

We might need a state object that can be passed through multiple levels:

```javascript
state = {
   completionCallbackRetrieved: false, //aka the current isAsync flag
   completionCallbackInvoked: false
};
```
And that reference gets passed to the two creation functions. Ideas?

Reply:
Good point. Added a test for this case.